### PR TITLE
Add IsBatteryLow condition node

### PR DESCRIPTION
**Pull request by naiveHobo**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1973 |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | Gazebo TurtleBot3 simulation |

---

## Description of contribution in a few bullet points

* Added `IsBatteryLow` BT condition node

## Description of documentation updates required from your changes

* Add new BT plugin to [navigation.ros.org](https://navigation.ros.org/)

### CODE REVIEW

File: nav2_behavior_tree/CMakeLists.txt [hunk 1]
```diff
@@ -89,6 +89,9 @@ list(APPEND plugin_libs nav2_distance_traveled_condition_bt_node)
 add_library(nav2_initial_pose_received_condition_bt_node SHARED plugins/condition/initial_pose_received_condition.cpp)
```
Comment (SteveMacenski) [inline nav2_behavior_tree/CMakeLists.txt:89]: Add to `md` and `website` in BT nodes and migration guide :smile:

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    const BT::NodeConfiguration & conf);
+
+  IsBatteryLowCondition() = delete;
+
+  BT::NodeStatus tick() override;
+
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<double>("min_battery", "Minimum battery voltage"),
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp:43]: Do you think this should be a % instead? I think most smart batteries will give you out N%, so we can check before N% than the voltage.

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      BT::InputPort<std::string>(
+        "battery_topic", std::string("/battery"), "Battery topic")
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp:45]: Lets do `battery_status` as the default

**Review by SteveMacenski — COMMENTED** [review]
Wow, that was quick, thanks for taking this on :+1:

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp (hunk 2 above)
Comment (naiveHobo) [reply nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp:43]: Yup, makes sense!

**Review by SteveMacenski — APPROVED** [review]
Add to MD and website and we can merge

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      BT::InputPort<std::string>(
+        "battery_topic", std::string("/battery_status"), "Battery topic")
+    };
+  }
+
+private:
+  void batteryCallback(std_msgs::msg::Float64::SharedPtr msg);
+
+  rclcpp::Node::SharedPtr node_;
+  rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr battery_sub_;
```
Comment (mikeferguson) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp:53]: why Float64? This is a deprecated topic in Foxy. It would be better to use standard message like [sensor_msgs::BatteryState](https://github.com/ros2/common_interfaces/blob/master/sensor_msgs/msg/BatteryState.msg) which exposes a percentage.

Comment (naiveHobo) [reply nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_battery_low_condition.hpp:53]: Ah great, I wasn't aware sensor_msgs::BatteryState exists.

**Comment by naiveHobo** [comment]
It looks like voltage is the only mandatory field in `sensor_msgs::BatteryState`:

https://github.com/ros2/common_interfaces/blob/3ee382893da4c34b43cf836d4c235d25ab4befab/sensor_msgs/msg/BatteryState.msg#L35

Should we still continue with percentage or is voltage a better option?

**Comment by mikeferguson** [comment]
> Should we still continue with percentage or is voltage a better option?

Perhaps both? could be parameterized.

**Comment by naiveHobo** [comment]
@mikeferguson looks okay now?

File: nav2_behavior_tree/CMakeLists.txt [hunk 5]
```diff
@@ -9,12 +9,12 @@ find_package(rclcpp_lifecycle REQUIRED)
 find_package(std_msgs REQUIRED)
 find_package(builtin_interfaces REQUIRED)
 find_package(geometry_msgs REQUIRED)
+find_package(sensor_msgs REQUIRED)
 find_package(nav2_msgs REQUIRED)
 find_package(nav_msgs REQUIRED)
 find_package(behaviortree_cpp_v3 REQUIRED)
 find_package(tf2_ros REQUIRED)
 find_package(tf2_geometry_msgs REQUIRED)
-find_package(std_msgs REQUIRED)
```
Comment (mikeferguson) [inline nav2_behavior_tree/CMakeLists.txt:17]: we still depend on std_msgs in this package (see L41 below), this line should be restored

**Review by mikeferguson — CHANGES_REQUESTED** [review]
We still need to find_package(std_msgs) - other than that, looks good.

File: nav2_behavior_tree/CMakeLists.txt (hunk 5 above)
Comment (naiveHobo) [reply nav2_behavior_tree/CMakeLists.txt:17]: There were two mentions of `std_msgs` on L9 and L17, just changed one to `sensor_msgs`.

Comment (mikeferguson) [reply nav2_behavior_tree/CMakeLists.txt]: I missed the second one!

**Review by mikeferguson — APPROVED** [review]

**Comment by SteveMacenski** [comment]
@mikeferguson don’t you think both is a little unclear? I’d prefer to work on just the % unless you think its atypical for implementations to make proper use of that field. % is a more linear than voltage and easier for a typical person to reason about when making robot navigation decisions (“I should probably redock at 20%” vs “I need to take a bunch of data and model what voltage I to threshold at”)

**Comment by mikeferguson** [comment]
I would have initially said "use percentage" - but it was pointed out that only the voltage is "required" in that message. For higher end robots, they should have a good estimate of battery state, but some lower end robots may not. I would suggest that maybe we invert the logic, so by default it uses percentage (since that is the better estimate), but there is a way to get a voltage-based mode?

**Comment by SteveMacenski** [comment]
> but there is a way to get a voltage-based mode?

I don't understand the question. Can you rephrase?

I agree that if we do either/or that percentage should be default. I could live with that.

**Comment by mikeferguson** [comment]
> I don't understand the question. Can you rephrase?

Not really a question. I was basically suggesting that we have two modes:

 * Default: uses percentage
 * There is some way to instead use voltage, for those robots that don't have a good percentage estimate.

**Review by SteveMacenski — COMMENTED** [review]
Needs markdown file in doc/parameters updated with new BT

**Review by SteveMacenski — APPROVED** [review]
